feat: Parse recipestep contents

ffaass opened the pull request:
Step에 대한 DTO를 만들어서 '["빵을 놓는다","고기를 올린다"]' 형식의 String을 List로 변환해서 View로 전달함
이후에 스텝 추가/수정시에도 JS에서 JSON string을 만들 필요 없이 JSON Array를 서버로 전달하면 DTO 재사용 가능

Resolves: #67

ffaass committed:
refactor: Rename RecipeDetailDTO to RecipeResponseDTO

- Apply vote results from slack channel

ffaass committed:
feat: Add RecipeStepDTO

ffaass committed:
feat: Include RecipeStepDTO into RecipeResponseDTO

- Change step model from RecipeStep to RecipeStepDTO in recipe detail
 view
- Modify template to show step contents
- Show recipe step contents by ordered list
- Modify dummy data

ffaass committed:
refactor: Extract step content converting logic

- Extract convert logic from dto to RecipeStepContentConverter
- Register content converter as Spring Bean
- Add tests
- String을 DTO 내부의 private method를 통해 List로 변환하고 있어서 테스트가 힘들었음
- DTO에 Jackson과의 의존 관계가 생겨버려서 추출함
- List -> String 간의 변환 로직은 현재 테스트에서만 사용하는데, 이후 스텝 추가/수정 때
 사용할 것 같아 남겨놓음

ffaass committed:
fix: Add ordered list in incompleted recipe view

CHANMYUNG reviewed (approved):
구현 잘 했어요. Approve합니다.

Lahavas reviewed (commented):
JSON과 String을 변환하는 동작에 대하여 Convert라는 단어와 Parse라는 단어가 혼용되는것같습니다. 구현에 대해서는 흠잡을 부분이 없네요 💯

ryuneeee commented on src/main/java/com/woowahan/techcamp/recipehub/recipestep/util/RecipeStepContentJsonConverter.java:
```diff
@@ -0,0 +1,44 @@
+import java.io.IOException;
+import java.util.List;
+
+public class RecipeStepContentJsonConverter implements RecipeStepContentConverter {
```
지금 단계에서는 무분별한 확장 설계로 보여요.
interface로 추상화 시키는건, Converter가 두개 이상일 시점에 다시 고민해봐도 될 듯 해요~

ryuneeee replied:
http://wonwoo.ml/index.php/post/995
자 Converter를 어-노테이션으로 만들어 봅시다! 
어차피 밑에서 리뷰한 것 처럼 Static Class로 만들 수 있다면, 고고

ffaass replied:
오호 이런게 있었네요!
이걸 쓰게 되면 entity 객체에서도 List로 들고 있을 수 있겠네요 ㅎㅎ 
그러면 DTO에서 별도의 변환 로직을 호출하지 않고도 원하던 기능을 구현할 수 있을 것 같네요!
오늘은 저녁까진 작업이 힘들고, 밤이나 내일쯤 바꿔서 다시 푸시해볼게요~

ffaass replied:
인터페이스로 뺀 이유가 있었는데 피드백받고 다시 생각해보니 저장되는 데이터에 버전이 생기는 경우 말고는 추상화할 이유가 딱히 없네요 ㅋㅋㅋㅋ 같이 수정해서 올릴게요!

ryuneeee commented on src/main/java/com/woowahan/techcamp/recipehub/recipestep/util/RecipeStepContentJsonConverter.java:
```diff
@@ -0,0 +1,44 @@
+        }
+
+        try {
+            return objectMapper.readValue(content, new TypeReference<List<String>>() {
```
이 클래스에서 상태를 가지고 있을 필요가 없을것 같은데, 클래스 안쪽에 method들을 다 static으로 만들어도 되지 않을까요?
예를 들어서 이런거죠.

ContentConverter.*toList*(content);

ryuneeee commented on src/main/java/com/woowahan/techcamp/recipehub/recipestep/util/RecipeStepContentJsonConverter.java:
```diff
@@ -0,0 +1,44 @@
+        }
+
+        try {
+            return objectMapper.writeValueAsString(contents);
```
이것도 마찬가지로

ContentConverter.*toContentString*(contents);

ryuneeee commented on src/test/java/com/woowahan/techcamp/recipehub/recipestep/dto/RecipeStepDTOTest.java:
```diff
@@ -0,0 +1,66 @@
+        assertThat(recipeStepDTO.getIngredients()).isEqualTo(recipeStep.getIngredients());
+        assertThat(recipeStepDTO.getSequence()).isEqualTo(recipeStep.getSequence());
+        assertThat(recipeStepDTO.getWriter()).isEqualTo(recipeStep.getWriter());
+        assertThat(recipeStepDTO.getRecipe()).isEqualTo(recipeStep.getRecipe());
```
이렇게 많이 비교하면 두 객체를 비교하는 method를 두개 중 하나의 클래스에 만들어도 될 것 같아요. 
왠지 만들어 두면 쓸 것 같은 느낌?

ffaass replied:
비교는 아직 이 테스트에서밖에 하고있지 않아서 메서드를 만들진 않았는데, equals로 비교 가능하게 하는건 어떻게 생각하시나요?

ryuneeee commented on src/test/java/com/woowahan/techcamp/recipehub/recipestep/util/RecipeStepContentJsonConverterTest.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Before
+    public void setUp() throws Exception {
+        contentList = Arrays.asList("식빵을 토스터기에 굽는다", "잼을 바른다");
```
Arrays.asList 굿👍

ryuneeee reviewed (changes requested):
Annotation 이용한 컨버터 만들어주세요~

ffaass committed:
Merge branch 'develop' into feature/parse-recipestep-contents

ffaass committed:
Merge branch 'develop' into feature/parse-recipestep-contents

ffaass committed:
refator: Change ContentConverter to JPA AttributeConverter

- Change RecipeStep content type from String to List
- Remove converter bean setting

Lahavas commented on src/main/java/com/woowahan/techcamp/recipehub/RecipehubConfig.java:
```diff
@@ -24,6 +24,7 @@ public AuthRequiredArgumentResolver authRequiredArgumentResolver() {
         return new AuthRequiredArgumentResolver();
     }
 
+
```
한줄 추가되어있네여

ryuneeee replied:
이거 너무 가혹한거 아닙니까 ㅋㅋㅋㅋㅋ 사람이 여유도 가지고 좀 그래야 ... 🎻🎺🏝

ryuneeee commented on src/main/java/com/woowahan/techcamp/recipehub/recipe/dto/RecipeResponseDTO.java:
```diff
@@ -38,19 +40,20 @@ public RecipeDetailDTO(Category category, String name, String imgUrl, List<Recip
+                .category(recipe.getCategory())
+                .imgUrl(recipe.getImgUrl())
+                .recipeSteps(
+                        stepStream.map(RecipeStepDTO::from).collect(Collectors.toList())
```
이거 collectors.toList mutable이라서, 혹시나 immutable로 바꿀 수 있다면 해주고, 아니면 너무 시간은 버리지 맙시다

ryuneeee reviewed (approved):
일단 저도 승인합니다 ~
